mbus-api/src/mbus_nats.rs

```diff
@@ -0,0 +1,337 @@
+                });
+            }
+            if log_error {
+                tracing::warn!(
+                    "Error while getting/creating stream '{}': {:?}. Retrying...",
+                    STREAM_NAME,
+                    err
+                );
+                log_error = false;
+            }
+            backoff_with_options(&mut tries, &options).await;
+        }
+    }
+
+    /// Returns subject for the message. Should be unique for each message.
+    fn subject(msg: &EventMessage) -> String {
+        format!("events.{}.{}", msg.category.to_string(), msg.metadata.id) // If category is volume
+                                                                           // and id is
+                                                                           // 'id', then the subject
+                                                                           // for the
+                                                                           // message is
+                                                                           // 'events.volume.id'
+    }
+
+    /// The payload for mbus publish from the message.
+    fn payload(msg: &EventMessage) -> BusResult<bytes::Bytes> {
+        Ok(Bytes::from(serde_json::to_vec(msg)?))
+    }
+}
+
+#[async_trait]
+impl Bus for NatsMessageBus {
+    /// Publish messages to the message bus atmost once. The message is discarded if there is any
+    /// connection error after some retries.
+    async fn publish(&mut self, message: &EventMessage) -> BusResult<u64> {
+        let options = BackoffOptions::publish_backoff_options();
+        let mut tries = 0;
+        let mut log_error = true;
+
+        let subject = NatsMessageBus::subject(message);
+        let payload = NatsMessageBus::payload(message)?;
+
+        loop {
+            let publish_request = self.jetstream.publish(subject.clone(), payload.clone());
+            let err = match publish_request.await {
+                Ok(x) => match x.await {
+                    Ok(y) => {
+                        return Ok(y.sequence);
+                    }
+                    Err(error) => error,
+                },
+                Err(error) => error,
+            };
+
+            if log_error {
+                tracing::warn!(
+                    "Error publishing message to jetstream: {:?}. Retrying...",
+                    err
+                );
+                let _stream = self.verify_stream_exists().await;
+                log_error = false;
+            }
+
+            if tries == options.max_retries {
+                return Err(Error::PublishError {
+                    retries: options.max_retries,
+                    payload: format!("{message:?}"),
+                    error: err.to_string(),
+                });
+            }
+            if let Some(error) = err.downcast_ref::<std::io::Error>() {
+                if error.kind() == ErrorKind::TimedOut {
+                    tries += 1;
```

Comment:
shouldn't we also bump tries if it's a different error, otherwise we could get stuck looping forever?

Reply:
For any other error, the backoff_with_options function is incrementing the tries count.